lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +55,136 @@ define((require, exports, module) ->
-        content = e
-      finally
-        popoverHandler.show($("#formula"), content)
-        editor.on("changeSelection", callbackHidePopover)
-        editor.session.on("changeScrollTop", callbackHidePopover)
-        editor.session.on("changeScrollLeft", callbackHidePopover)
-        return
 
-    callback = (editor) -> createPopover(editor)
+      initPopover: ->
+        cursorRow = editor.getCursorPosition().row
+        [ch.curStart, ch.curEnd] = ch.getEquationRange(cursorRow)
+        popoverPosition = ch.getPopoverPosition(ch.curEnd)
+        try
+          content = ch.getCurrentFormula()
+        catch e
+          content = e
+        finally
+          popoverHandler.show(jqFormula(), content, popoverPosition)
 
-    destroyPopover = -> popoverHandler.hide($("#formula"))
+      updatePopover: ->
+        try
+          content = ch.getCurrentFormula()
+        catch e
+          content = e
+        finally
+          popoverHandler.setContent(jqFormula(), content)
 
-    createPopover = (editor) ->
-      unless katex?
-        initKaTeX(onLoaded)
+      updateCallback: ->
+        if ch.lastChangeTime?
+          curTime = Date.now()
+          ch.currentDelayedUpdateId = setTimeout(ch.updateCallback, ch.updateDelay - (Date.now() - ch.lastChangeTime))
+          ch.lastChangeTime = null
+        else
+          ch.updatePopover()
+          ch.currentDelayedUpdateId = null
+
+      delayedUpdatePopover: ->
+        if ch.currentDelayedUpdateId?
+          ch.lastChangeTime = Date.now()
+          return
+        ch.currentDelayedUpdateId = setTimeout(ch.updateCallback, ch.updateDelay)
+
+      updatePosition: ->
+        popoverHandler.setPosition(jqFormula(), ch.getPopoverPosition(ch.curEnd))
+
+      handleCurrentContext: ->
+        currentContext = latexContextParser.getContext(editor.session, editor.getCursorPosition().row)
+        if not ch.contextPreviewExists and currentContext == "equation"
+          ch.contextPreviewExists = true
+          if not katex?
+            initKaTeX(ch.initPopover)
+          else
+            ch.initPopover()
+          editor.on("change", ch.delayedUpdatePopover)
+          editor.session.on("changeScrollTop", ch.updatePosition)
+        else if ch.contextPreviewExists and currentContext != "equation"
+          ch.contextPreviewExists = false
+          editor.off("change", ch.delayedUpdatePopover)
+          editor.session.off("changeScrollTop", ch.updatePosition)
+          popoverHandler.destroy(jqFormula())
+    }
+
+    sh = selectionHandler = {
+      hideSelectionPopover: ->
+        popoverHandler.destroy(jqFormula())
+        editor.off("changeSelection", sh.hideSelectionPopover)
+        editor.session.off("changeScrollTop", sh.hideSelectionPopover)
+        editor.session.off("changeScrollLeft", sh.hideSelectionPopover)
         return
-      onLoaded()
+
+      renderSelectionUnderCursor: ->
+        try
+          cursorPosition = $("textarea.ace_text-input").position()
```

Comment:
any sense to use `renderer.textToScreenCoordinates` from the cursor text position?